### ISSUE TITLE
Error in rsqlite_send_query(conn@ptr, statement) : database is locked

### ISSUE THREAD

**Issue by HenrikBengtsson** [r-lib/revdepcheck#107]
While running:
```r
revdep_check(bioc = TRUE, num_workers = 32L,
	     timeout = as.difftime(30, units = "mins"), quiet = FALSE)
```
in batch mode, R terminated with the following output:
```r
<U+221A> bigstep 0.7.4                          <U+2500><U+2500> E: 0     | W: 0     | N: 0    
<U+221A> DupChecker 1.14.0                      <U+2500><U+2500> E: 1     | W: 0     | N: 1    
<U+221A> acc 1.3.3                              <U+2500><U+2500> E: 0     | W: 0     | N: 1    
Warning: Couldn't set synchronous mode: database is locked
Use `synchronous` = NULL to turn off this warning.
Error in rsqlite_send_query(conn@ptr, statement) : database is locked
Calls: revdep_check ... initialize -> initialize -> rsqlite_send_query -> .Call
31: (function () 
    {
        traceback(2L)
        dump.frames(dumpto = getOption("startup.session.dumpto", 
            "last.dump"), to.file = TRUE)
    })()
30: stop(list(message = "database is locked", call = rsqlite_send_query(conn@ptr, 
        statement), cppstack = list(file = "", line = -1L, stack = c("/cbc/henrik/R/x86_64-pc-linux-gnu-library/3.4/RSQLite/libs/RSQLite.so(Rcpp::exception::exception(
char const*, bool)+0x5d) [0x2b10ee1ed97d]", 
    "/cbc/henrik/R/x86_64-pc-linux-gnu-library/3.4/RSQLite/libs/RSQLite.so(Rcpp::stop(std::string const&)+0x27) [0x2b10ee1eda97]", 
    "/cbc/henrik/R/x86_64-pc-linux-gnu-library/3.4/RSQLite/libs/RSQLite.so(SqliteResultImpl::raise_sqlite_exception(sqlite3*)+0x27) [0x2b10ee202237]", 
    "/cbc/henrik/R/x86_64-pc-linux-gnu-library/3.4/RSQLite/libs/RSQLite.so(SqliteResultImpl::raise_sqlite_exception() const+0) [0x2b10ee2022b0]", 
    "/cbc/henrik/R/x86_64-pc-linux-gnu-library/3.4/RSQLite/libs/RSQLite.so(SqliteResultImpl::SqliteResultImpl(sqlite3*, std::string const&)+0x26) [0x2b10ee204466]", 
    "/cbc/henrik/R/x86_64-pc-linux-gnu-library/3.4/RSQLite/libs/RSQLite.so(SqliteResult::SqliteResult(boost::shared_ptr<SqliteConnection> const&, std::string const&)+0
x4d) [0x2b10ee2005cd]", 
    "/cbc/henrik/R/x86_64-pc-linux-gnu-library/3.4/RSQLite/libs/RSQLite.so(rsqlite_send_query(Rcpp::XPtr<boost::shared_ptr<SqliteConnection>, Rcpp::PreserveStorage, &(
void Rcpp::standard_delete_finalizer<boost::shared_ptr<SqliteConnection> >(boost::shared_ptr<SqliteConnection>*)), false> const&, std::string const&)+0x3f) [0x2b10ee20
83ef]", 
    "/cbc/henrik/R/x86_64-pc-linux-gnu-library/3.4/RSQLite/libs/RSQLite.so(RSQLite_rsqlite_send_query+0x73) [0x2b10ee1f2ef3]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4865b6]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4916b9]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x7c3) [0x4ca813]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4ce67b]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x5e2) [0x4ca632]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cc6e5]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_applyClosure+0x26f) [0x4cc435]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4da5bf]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x206) [0x4ca256]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4c9ff1]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x406) [0x4ca456]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4c9ff1]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x406) [0x4ca456]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4d98ef]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x206) [0x4ca256]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cc6e5]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(R_execMethod+0x3b6) [0x4cd08f]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/library/methods/libs/methods.so(+0x584d) [0x2b10ececd84d]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x523fdd]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4da462]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x206) [0x4ca256]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cc6e5]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_applyClosure+0x26f) [0x4cc435]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4da5bf]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x206) [0x4ca256]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cc6e5]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_applyClosure+0x26f) [0x4cc435]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4da5bf]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x206) [0x4ca256]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cc6e5]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_applyClosure+0x26f) [0x4cc435]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x8c6) [0x4ca916]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4ce67b]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x5e2) [0x4ca632]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cc6e5]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(R_execMethod+0x3b6) [0x4cd08f]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/library/methods/libs/methods.so(+0x584d) [0x2b10ececd84d]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x523fdd]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x81d) [0x4ca86d]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cf56d]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x5e2) [0x4ca632]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4ce67b]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x5e2) [0x4ca632]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cc6e5]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_applyClosure+0x26f) [0x4cc435]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x8c6) [0x4ca916]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cf56d]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x5e2) [0x4ca632]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4ce67b]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x5e2) [0x4ca632]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cc6e5]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_applyClosure+0x26f) [0x4cc435]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x8c6) [0x4ca916]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4ce67b]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x5e2) [0x4ca632]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cc6e5]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(R_execMethod+0x3b6) [0x4cd08f]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/library/methods/libs/methods.so(+0x584d) [0x2b10ececd84d]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x523fdd]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x81d) [0x4ca86d]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cc6e5]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_applyClosure+0x26f) [0x4cc435]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x8c6) [0x4ca916]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4ce67b]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x5e2) [0x4ca632]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4d04a3]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4da928]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x206) [0x4ca256]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cc6e5]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_applyClosure+0x26f) [0x4cc435]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x8c6) [0x4ca916]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4ce67b]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x5e2) [0x4ca632]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cc6e5]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_applyClosure+0x26f) [0x4cc435]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x8c6) [0x4ca916]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4ce67b]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x5e2) [0x4ca632]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cc6e5]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_applyClosure+0x26f) [0x4cc435]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4da5bf]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x206) [0x4ca256]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cc6e5]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_applyClosure+0x26f) [0x4cc435]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x8c6) [0x4ca916]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4ce67b]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x5e2) [0x4ca632]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4cc6e5]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_applyClosure+0x26f) [0x4cc435]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R() [0x4da5bf]", 
    "/home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/bin/exec/R(Rf_eval+0x206) [0x4ca256]"
    ))))
29: .Call(RSQLite_rsqlite_send_query, con, sql)
28: rsqlite_send_query(conn@ptr, statement)
27: initialize(value, ...)
26: initialize(value, ...)
25: new("SQLiteResult", sql = statement, ptr = rsqlite_send_query(conn@ptr, 
        statement), conn = conn)
24: .local(conn, statement, ...)
23: dbSendQuery(conn, statement, ...)
22: dbSendQuery(conn, statement, ...)
21: .local(conn, statement, ...)
20: dbGetQuery(db, "SELECT * FROM packages ORDER BY Package, Version")
19: dbGetQuery(db, "SELECT * FROM packages ORDER BY Package, Version")
18: eval(substitute(expr), envir = list(db = con), enclos = parent.frame())
17: eval(substitute(expr), envir = list(db = con), enclos = parent.frame())
16: with_db(db_file, {
        dbGetQuery(db, "SELECT * FROM packages ORDER BY Package, Version")
    })
15: db_all_packages(db_file)
14: write_packages_files(dir, db_file)
13: create_empty_PACKAGES(dir)
12: create_cache_if_needed()
11: get_cached_repos(type)
10: get_crancache_repos(type = mytype)
9: crancache::download_packages(pkgname, dir, repos = get_repos(bioc = TRUE), 
       quiet = TRUE)
8: force(code)
7: with_envvar(c(CRANCACHE_REPOS = "cran,bioc", CRANCACHE_QUIET = "yes"), 
       crancache::download_packages(pkgname, dir, repos = get_repos(bioc = TRUE), 
           quiet = TRUE)[, 2])
6: check_proc(pkgdir, pkgname, version = version)
5: check_task(state, task)
4: do_task(state, task)
3: run_event_loop(state)
2: revdep_run(pkg, quiet = quiet, timeout = timeout, num_workers = num_workers)
1: revdep_check(bioc = TRUE, num_workers = availableCores(), timeout = as.difftime(30, 
       units = "mins"), quiet = FALSE)
```

## Session information

```r
> library("revdepcheck")
> sessionInfo()
R version 3.4.2 (2017-09-28)
Platform: x86_64-pc-linux-gnu (64-bit)

Matrix products: default
BLAS: /home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/lib/libRblas.so
LAPACK: /home/shared/cbc/software_cbc/R/R-3.4.2-20170928/lib64/R/lib/libRlapack.so

locale:
[1] C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] revdepcheck_1.0.0.9000

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.13           compiler_3.4.2         prettyunits_1.0.2     
 [4] base64enc_0.1-4        remotes_1.1.0          tools_3.4.2           
 [7] progress_1.1.2.9000    digest_0.6.12          bit_1.1-12            
[10] jsonlite_1.5           RSQLite_2.0            memoise_1.1.0         
[13] tibble_1.3.4           debugme_1.0.2          rlang_0.1.2           
[16] whoami_1.1.1           DBI_0.7                yaml_2.1.14           
[19] curl_3.0               gmailr_0.7.1           withr_2.0.0           
[22] httr_1.3.1             knitr_1.17             desc_1.1.1            
[25] rappdirs_0.3.1         hms_0.3                rprojroot_1.2         
[28] bit64_0.9-7            glue_1.1.1             crancache_0.0.0.9000  
[31] parsedate_1.1.3        R6_2.2.2               processx_2.0.0.1      
[34] rcmdcheck_1.2.1.9002   sessioninfo_1.0.1.9000 callr_1.0.0.9000      
[37] blob_1.1.0             magrittr_1.5           backports_1.1.1       
[40] boxes_0.0.0.9000       clisymbols_1.2.0       assertthat_0.2.0      
[43] cranlike_1.0.1         crayon_1.3.4
```

**Comment by gaborcsardi** [comment]
Was this a one off thing, or you can reproduce it?

**Comment by HenrikBengtsson** [comment]
I think (but not 100%) sure this was the first time I seen this one.  I'm sharing this here so it gets indexed in case someone else run into it as well.  I'll report back here if I see it again.

A question related to this: when running in parallel, is there more than one process that writes to the SQLite database, or is that all handled by the master R process?  Although they put lots of efforts in concurrency/locking, I don't think SQLite is bullet proof and it depends lots on what file system you use etc.  Disclaimer: It's been years since I looked into that so my conclusions might be outdated.

**Comment by HenrikBengtsson** [comment]
If it adds any information, when restarting `revdep_check()`, it successfully picked up where it was terminated.

**Comment by gaborcsardi** [comment]
Only the master process is writing to the DB. In theory you could run multiple checks using the same DB, but that's probably a bad idea. Maybe you used another R process at the same time, to query the results? That should be OK, though, because it only reads, and only makes short queries. But I'll add a timeout to the db open, anyway.

**Comment by HenrikBengtsson** [comment]
Ok.  Modulo my occasional absent mindedness, I'm pretty sure no other process was accessing this DB at the same time.

**Comment by gaborcsardi** [comment]
Looks like this is actually coming from `cranlike` which was not entirely parallel-safe, until today. So this should never happen again.
